spec/arrays.dd

```diff
@@ -24,11 +24,31 @@ int* p;
 ---------
 )
 
-        $(P These are simple pointers to data, analogous to C pointers.
+        $(P A pointer to type $(I T) has a value which is a reference (address) to another
+        object of type $(I T). It is commonly called a $(I pointer to T).
+        )
+
+        $(P If a pointer contains a $(I null) value, it is not pointing to a valid object.)
+
+        $(P A pointer to $(I T) must either contain a $(I null) value,
+        or point to a valid object of type $(I T).)
+
+        $(IMPLEMENTATION_DEFINED
+        $(OL
+        $(LI The behavior when a $(I null) pointer is dereferenced. Typically the program
+        will be aborted.)
+        ))
+
+        $(UNDEFINED_BEHAVIOR
+        $(OL
+        $(LI Dereferencing a pointer that is not $(I null) and does not point
+        to a valid object of type $(I T).)
+        ))
+
+        $(BEST_PRACTICE These are simple pointers to data.
         Pointers are provided for interfacing with C and for
         specialized systems work.
-        There
-        is no length associated with it, and so there is no way for the
+        There is no length associated with it, and so there is no way for the
         compiler or runtime to do bounds checking, etc., on it.
         Most conventional uses for pointers can be replaced with
         dynamic arrays, $(D out) and $(D ref) parameters,
```

Comment:
The elephant in the room is that `key in aa` for associative arrays returns a pointer. In that case, a pointer is not only necessary (indicates the key doesn't exist in the AA), but also perfectly safe (if it's not null, it always points to a valid value in the AA).  This text almost makes it sound like `key in aa` should be avoided, or is somehow "specialized systems work".

Reply:
Unfortunately, the `key in aa` is dangerous if not used carefully. For example, the value can be deleted from the aa and yet the pointer remains. It's a poor design.

Reply:
Fair enough.